### ISSUE TITLE
QGIS 3.22.8 'Create postgres provider' causing very slow project opening times

### ISSUE THREAD

**Issue by weca-theo** [qgis/QGIS#51110]
### What is the bug or the crash?

With QGIS 3.x, opening a .qgz or .qgs project file containing around 100 remote PostGIS layers takes around 2 minutes to load (1 minute of the blue loading bar to reach 100%, then another minute hanging on 100% until the map appears on screen). More layers (another 50) are to be added to this QGIS file, so project load times will only increase. This is an unacceptable waiting time for mass rollout to 50+ novice GIS users within my organisation.

All PostGIS layers are hosted in a remote Azure PostgreSQL Flexible Server general purpose instance, located in the same geographic region as our QGIS users. All PostGIS layers have a primary key and spatial index. Most layers are unchecked by default, with only a basemap and a boundary layer visible on project load. Rendering all PostGIS data on a map is fast/acceptable, it's the initial QGIS _project_ load time that is the problem. I've manually injected `estimatedmetadata=true` into the .qgs file for each data source (this is meant to speed project load up) and I've checked the 'Trust project when data source has no metadata' option in Project>Properties>Data Sources tab (the tooltip for this says 'Speed up project loading by skipping data checks in PostgreSQL layers.'

In the QGIS debugging panel (View>Panels>Debugging/Development Tools) then selecting the 'Profiler' tab, the 'Startup' Category drop down only shows 2 or 3 seconds of loading, however the 'Project Load' drop down shows that the 2 minute loading time is due to 'Reading map layers'. Each layer takes between 1 and 2 seconds to load, **with almost 99% of that time due to `Create postgres provider`**. Longest loading layer takes almost 2 seconds and _only_ has 700,000 point features. Example of layer load shown in screenshot.

Testing further, an identical dataset loaded from localhost is actually slower than loaded from Azure (dataset dependant probably), however, curiously there is minimal time spent (relatively) on `Create postgres provider`.

![image](https://user-images.githubusercontent.com/83225776/205524317-10a5e411-8b9d-401a-9112-dce8acda93c5.png)

### Steps to reproduce the issue

1. create a new QGIS project file
2. make sure 'estimate metadata' is turned on when creating PostgreSQL connection
3. add multiple large (0.5-1 million row) PostGIS layers via 'Add PostGIS Layers' option
4. save file.
5. close and open file
6. observe 'Create postgres provider' time in the debug panel. Each layer has a delay here.

### Versions

Tested with versions:
3.22.8
3.22.5
3.16.6

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

Using QGIS on corporate laptop, connected into corporate VPN. Network performance does degrade due to this enforced setup, but not significantly so.

**Comment by nyalldawson** [comment]
I suggest re-testing with 3.28 and using the new "SQL Query logger" tool in the developer tool panel to diagnose which queries are taking the time to execute during layer loading. That may give some clues on possible optimisations on your db.

**Comment by weca-theo** [comment]
> I suggest re-testing with 3.28 and using the new "SQL Query logger" tool in the developer tool panel to diagnose which queries are taking the time to execute during layer loading. That may give some clues on possible optimisations on your db.

we only use the LTR stable QGIS release (IT policy) which is version 3.22. We can install (or test) newer releases since then. I also can't test outside of our corporate network, as I need to be inside the corporate VPN in order to connect to the databases.

I will have a look at the Azure PostgreSQL logs though in Azure portal, specifically looking at the queries and query durations during a QGIS project load. I can report back on this within 24 hours.

**Comment by weca-theo** [comment]
I've managed to download QGIS 3.28.1 on another company machine, where I'm able to install non-LTR releases for testing. Using the new SQL Query Logger debug tool, I can see over 1000 SQL queries running on project load. Each one takes between 15ms and 50ms (these will be slightly faster on the machine I'm now testing on, as it's closer to our VPN gateway to Azure). This log explains the 1-2minute initial load time. I can attach the full .json log here, but snippets look like (pardon the poor formatting Github applies):


  `{

    "Initiator": "QgsPostgresProvider",

    "Location": "src\\providers\\postgres\\qgspostgresprovider.cpp:1642 (QgsPostgresProvider::hasSufficientPermsAndCapabilities)",

    "Provider": "postgres",

    "Row count": "0",

    "SQL": "SELECT 1 FROM pg_class,pg_namespace WHERE pg_class.relnamespace=pg_namespace.oid AND pg_has_role(relowner,'MEMBER') AND relname='rider_zones_27700' AND nspname='travelwest'",

    "Started at": "2022-12-05T11:58:08.844",

    "Total time (ms)": "32",

    "URI": "dbname='db_name' host=server-name.postgres.database.azure.com port=5432 user='username' password='xxxx'"
  },`

Each JSON object (1000+ of these) contains `"Provider": "postgres"` and `"Initiator": "QgsPostgresProvider"` (with a handful of initiators being `"unknown"` or `"Initiator": "QgsPostgresConn"`. The provider location can vary from:
`"Location": "src\\providers\\postgres\\qgspostgresprovider.cpp:4137 (QgsPostgresProvider::getGeometryDetails)"`
`"Location": "src\\providers\\postgres\\qgspostgresprovider.cpp:218 (QgsPostgresProvider::QgsPostgresProvider)"`
`"Location": "src\\providers\\postgres\\qgspostgresprovider.cpp:986 (QgsPostgresProvider::loadFields)"`
`"Location": "src\\providers\\postgres\\qgspostgresprovider.cpp:72 (tableExists)"`
`"Location": "src\\providers\\postgres\\qgspostgresprovider.cpp:1731 (QgsPostgresProvider::determinePrimaryKey)"`
`"Location": "src\\providers\\postgres\\qgspostgresconn.cpp:1268 (QgsPostgresConn::resetSessionRole)"`

So many trips to determine Postgres settings and configurations. Seems bloated.

**Comment by strk** [comment]
I'm often surprised myself by the many round trips.  I suspect the origin of the problem is in some misuse of connection pools / caches. See for example #47391

**Comment by weca-theo** [comment]
Regardless of whether the layer is loaded from:
-DB Manager plugin
-Add Vector Layer

And regardless if the layer is:
-a direct connection to the table
-a VIEW of the table
-a MATERIALIZED VIEW of the table

The issue remains. QGIS make unnecessary calls to PostGIS tables on project load. A .qgz file with 100+ layers can take minutes to load.

**Comment by borysiasty** [comment]
All I was doing for #38114 was optimizing the slowest query, what fixed most of the problem for me (1 hour -> 1 minute speedup), so I didn't dig into those unnecessary queries. Further improvements would be great of course.

**Comment by weca-theo** [comment]
I've written this up on GIS SE if anyone is interested (perhaps more concise than I have here):
https://gis.stackexchange.com/questions/448015/long-qgis-project-loading-time-due-to-qgspostgresprovider-sql-calls

**Comment by weca-theo** [comment]
I've also discovered that the SQL queries viewed in the SQL debugger make up around 20 seconds of the total 2 minute project load time. Therefore the majority of the project load time is either latency or whatever else is running in `Create postgres provider`. I've also found that number of columns in a layer's Postgres table, does not have an affect on `Create postgres provider` load time (I tested with a 1mil row table of 20 columns, then a duplicate table with only the PK and geometry column remaining. Timings were similar).

Just looping in some other who may be able to provide insight to this issue:
@palmerj @gioman @troopa81

**Comment by weca-theo** [comment]
> I would say ~100 remote layers is not a good idea with a remote database due to round trip latency on each SQL query. QGIS is a very chatty app with backend database APIs. Can you trace how many SQL queries are sent per layer creation? I would guess it's many 10s of queries.

As mentioned above, for each layer in the QGIS file: each column found in the layer's table will be queried with these 2 queries:
`"SQL": "SELECT domain_name, domain_schema FROM information_schema.columns WHERE table_name='example_table_name' AND column_name='example_column_name'"`
and
`"SQL": "SELECT typtype FROM pg_type WHERE typname='example_data_type'"`

So if a table has 20 columns, this means 40 SQL queries for that table (layer).

The odd thing is, some layers aren't queried at all! I can't figure out why.

**Comment by weca-theo** [comment]
> > The odd thing is, some layers aren't queried at all! I can't figure out why.
> 
> Maybe some tables are not in geometry_columns view? Can you see if they are spatial or not?

All PostgreSQL layers are spatial, bar 1. There are also a couple of WMS layers, but they aren't the issue.

**Comment by weca-theo** [comment]
> All showing up in geometry_columns?
> 
> Regardless of this, the PostgreSQL drivers is known to be chatty. It's likely some more significant development would could be done to improve that. To help that you should try to better detail the layers you have in your project and then a trace of the SQL calls for creation of the postgresql provider instances.

Yes, all layers appearing in the QGIS SQL trace are present in the geometry_columns view. I'll attach the raw SQL trace to this message, with sensitive info changed.
[qgis3_28_1_SQL_trace_230105.zip](https://github.com/qgis/QGIS/files/10353235/qgis3_28_1_SQL_trace_230105.zip)

**Comment by strk** [comment]
Feedback was given so I removed the label, and the "stale" one as well. This is a confirmed bug.

**Comment by weca-theo** [comment]
Thanks @strk . Will this be assigned to a dev? I'm not familiar with the process.

To summarise my findings:

- QGIS slow project opening time due to excessive chat between client and remote PostgreSQL/PostGIS server hosted in Azure
- 2-3 minute loading time for QGIS 3.22.8 project file containing around 100 PostGIS layers
- all PostGIS tables have spatial indexes
- all PostGIS tables have primary keys
- all PostGIS tables have same projection (British National Grid: 27700)
- all PostGIS tables are under 1 million rows each. In fact, 90% of tables are under 1000 features each.
- QGIS set to estimate metadata for every layer
- QGIS set to not check PK unicity for every layer
- QGIS SQL trace (had to use newer QGIS release to use this tool) showing 15-50ms request for **every** table column, but for only 50% of the total loaded tables.
- Why are 50% of the tables not queried on load? I can't see any pattern. Why can't _all_ tables skip this SQL query stage?
- the bulk of the SQL queries on each column are:

`"SQL": "SELECT domain_name, domain_schema FROM information_schema.columns WHERE table_name='example_table_name' AND column_name='example_column_name'"`
and
`"SQL": "SELECT typtype FROM pg_type WHERE typname='example_data_type'"`

I know there latency is expected between a client and a remote Azure PostgreSQL server, but the main issue here is why _every column_ in ~half of the tables is being queried on project opening. Can we not simply cut this querying down or out entirely?

**Comment by troopa81** [comment]
I'll try to take a look about this one if nobody steps in, before mid february I hope.

**Comment by weca-theo** [comment]
thanks @troopa81 , it would be good to understand why some PostGIS table types spawn SQL queries on load, and why some do not at all.

**Comment by troopa81** [comment]
I take a look about this issue, and I figure out that trust parameter was not used in recent version (since 3.28). I proposed a fix in #51854 for this one

But it won't change the main part of your issue, which is that we fire many requests to the database in order to get metadata (fields, types, geometry type...).

We could improve cache [here](https://github.com/qgis/QGIS/blob/master/src/providers/postgres/qgspostgresprovider.cpp#L2150) to avoid retrieve info for the same type many times (236 times for varchar for instance, so 8s in total for your project) or we can group requests [here](https://github.com/qgis/QGIS/blob/master/src/providers/postgres/qgspostgresconn.cpp#L1164).

It would improve things a bit but I think the right and significant approach is this one described in QEP #261, meaning a broader trust parameter to avoid reloading metadata any time you open your project.

Both approach could be complimentary, but that's a lot of work and would require funding.

**Comment by weca-theo** [comment]
@troopa81 thanks. I'll keep an eye on the QEP [#261](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/261) for now.

**Comment by teo113** [comment]
@troopa81 do you think this is related?

https://github.com/qgis/QGIS/pull/53112

**Comment by weca-theo** [comment]
@wonder-sk hi, I've seen you're work on https://github.com/qgis/QGIS/pull/53112 . Would the issue above be resolved by your fix? thanks

**Comment by wonder-sk** [comment]
@weca-theo yeah my fix in #53112 helps for PostgreSQL layers where columns have no edit widget type defined (`<editWidget type="">`). 

Also #53069 from my colleague should help in your case.

**Comment by weca-theo** [comment]
> > do you know which version of QGIS both these fixes will appear in?
> 
> QGIS 3.32 (out in June)

As seen on other threads, QGIS 3.32.0 experiencing no noticeable speed up of project load, with regressions noted. Your fix https://github.com/qgis/QGIS/pull/53112 certainly seems to remove the SQL queries on layers with `<editWidget type="">`, however the project load time has remained the same due to excessive SQL calls to the database, on numerous calls such as:

`SELECT postgis_version()` (the PostGIS version should only be called once surely? If the data is in the same database).

`SELECT postgis_geos_version(), postgis_proj_version()`

`SELECT has_schema_privilege`

`SELECT has_table_privilege`
